third_party/move/move-vm/runtime/src/runtime_type_checks.rs

```diff
@@ -0,0 +1,683 @@
+// Copyright © Aptos Foundation
+// SPDX-License-Identifier: Apache-2.0
+
+use crate::{frame_type_cache::FrameTypeCache, interpreter::Stack, loader::Resolver};
+use move_binary_format::{
+    errors::*,
+    file_format::{Ability, AbilitySet, Bytecode},
+};
+use move_core_types::vm_status::StatusCode;
+use move_vm_types::{loaded_data::runtime_types::Type, values::Locals};
+
+pub(crate) trait RuntimeTypeCheck {
+    fn pre_execution_type_stack_transition(
+        local_tys: &[Type],
+        locals: &Locals,
+        _ty_args: &[Type],
+        _resolver: &Resolver,
+        operand_stack: &mut Stack,
+        instruction: &Bytecode,
+    ) -> PartialVMResult<()>;
+
+    fn post_execution_type_stack_transition(
+        local_tys: &[Type],
+        ty_args: &[Type],
+        resolver: &Resolver,
+        operand_stack: &mut Stack,
+        ty_cache: &mut FrameTypeCache,
+        instruction: &Bytecode,
+    ) -> PartialVMResult<()>;
+}
+
+/// Paranoid type checks to perform before instruction execution.
```

Comment:
If this is meant for the module, you can do `//! Your description here` on top of the file? This also seems like a good documentation to add to trait definition?